token_provider.go

```diff
@@ -0,0 +1,82 @@
+package incognia
+
+import (
+	"errors"
+	"time"
+)
+
+type TokenProvider interface {
+	GetToken() (Token, error)
+}
+
+type Token interface {
+	IsExpired() bool
+	GetExpiresAt() time.Duration
+}
+
+var (
+	ErrTokenNotFound = errors.New("token not found in memory")
+	ErrTokenExpired  = errors.New("incognia token expired")
+)
+
+type ManualRefreshTokenProvider struct {
+	tokenClient *TokenClient
+	token       Token
+}
+
+func NewManualRefreshTokenProvider(tokenClient *TokenClient) *ManualRefreshTokenProvider {
+	return &ManualRefreshTokenProvider{tokenClient: tokenClient}
+}
+
+func (t *ManualRefreshTokenProvider) GetToken() (Token, error) {
+	if t.token != nil && !t.token.IsExpired() {
+		return t.token, nil
+	}
+
+	if t.token.IsExpired() {
+		return nil, ErrTokenExpired
+	}
+
+	return nil, ErrTokenNotFound
+}
```

Comment:
You're checking if the token is expired before the `t.token == nil` check (last return statement), which could cause a null pointer exception. I'd rewrite the method like this:

```suggestion
func (t *ManualRefreshTokenProvider) GetToken() (Token, error) {
	if t.token == nil {
		return nil, ErrTokenNotFound
	}
	if t.token.IsExpired() {
		return nil, ErrTokenExpired
	}
	return t.token, nil
}
```